Add HA Harbor deployment (#24)
SovereignCloudStack/k8s-harbor: README.md

Also, usually flux2 resources are used for e.g. Helm, so flux2 controllers need to be installed in any destination cluster.

Repository structure:
- `base`
  - Contains Harbor base configuration and base helm release definition
  - Can be deployed as follows `kubectl apply -k base/`
    - In this case harbor is deployed via clusterIP and without additional services and persistence
    - Run `kubectl port-forward svc/harbor 8080:80` and access harbor at http://localhost:8080
  - Can be referenced by kustomize overlays, see e.g. `envs/public/`
    - To override base config use *harbor-config* configmap, see e.g. [harbor-config.yaml](envs/ci-testing/harbor-config.yaml)
- `operators`
  - Contains helm release definitions for cert-manager, ingress-nginx, postgres-operator and redis-operator
  - All operators can be deployed at once using `kubectl apply -k operators/`
  - Separate operators can be deployed using e.g. `kubectl apply -k operators/redis/`
- `postgres`
  - Contains CR `postgresql` - postgresql cluster with basic configuration

>   - Postgres-operator has to be installed first, and then it can be installed by `kubectl apply -k postgres/`

Reviewer: ```suggestion
  - Postgres-operator has to be installed first, and then postgresql cluster can be deployed by `kubectl apply -k postgres/`
```

- `redis`
  - Contains CR `RedisFailover` - redis sentinel with basic configuration

>   - Redis-operator has to be installed first, and then it can be installed by `kubectl apply -k redis/`

Reviewer: ```suggestion
  - Redis-operator has to be installed first, and then redis sentinel cluster can be deployed by `kubectl apply -k redis/`
```

  - Each subdirectory define deployment of Harbor
    - refers to `base`
    - adds *harbor-config* configmap
    - It can contain patches and other kustomizations, like `envs/public-ha/redis/`

## Installation

### Prerequisites

#### Kubernetes v1.20+
```
export KUBECONFIG=/path/to/kubeconfig
```

#### FluxCD
```
curl -s https://fluxcd.io/install.sh | sudo FLUX_VERSION=0.40.2 bash
flux install
```

### Example public HA environment installation

#### Install and wait for operators
```
$ kubectl apply -k operators/
$ flux get helmrelease -n default
NAME                    REVISION        SUSPENDED       READY   MESSAGE
cert-manager            v1.11.0         False           True    Release reconciliation succeeded
ingress-nginx           4.5.2           False           True    Release reconciliation succeeded
postgres-operator       1.9.0           False           True    Release reconciliation succeeded
redis-operator          3.2.7           False           True    Release reconciliation succeeded
```

> Note: Install separate operators by e.g.:
> ```
> kubectl apply -k operators/redis/
> kubectl apply -k operators/postgres/
> ```

> #### Create redis and postgres cluster

Reviewer: ```suggestion
#### Create redis and postgres clusters
```

```
bash envs/public-ha/redis/redis-secret.bash
kubectl apply -k envs/public-ha/redis/
kubectl apply -k envs/public-ha/postgres/
```

#### Install Harbor

- Take *ingress-nginx-controller* LoadBalancer IP address and create DNS record for Harbor.

- Generate secrets and install Harbor:
  ```
  bash base/harbor-secrets.bash
  bash envs/public-ha/swift-secret.bash <username> <password>
  kubectl apply -k envs/public-ha/
  ```

#### All in one installation using FluxCD Kustomization and GitRepository reconciliation

```
bash envs/public-ha/redis/redis-secret.bash
bash base/harbor-secrets.bash
bash envs/public-ha/swift-secret.bash <username> <password>
# --branch/tag can be specified, default to master
flux create source git k8s-harbor --url=https://github.com/SovereignCloudStack/k8s-harbor --interval=5m
kubectl apply -f envs/public-ha/public-ha.yaml
```

> ### Public HA environment threat model

Reviewer: ```suggestion
### Public environment threat model
```

We define the threat model to generally trust the network. It is mainly based on the fact, that all the services live
in the same k8s cluster, so services can communicate with each other without certificate verification because

> we do not expect MITM attacks in the Kubernetes private network. We use HA databases which are external from

Reviewer: ```suggestion
we do not expect MITM attacks in the Kubernetes private network. In the case of HA databases (`envs/public-ha`), they are external to Harbor
```

> Harbor's point of view, but still running in the same k8s cluster, so just basic auth is implemented(it is easy to do).

Reviewer: ```suggestion
, but they are still running in the same k8s cluster, so just basic auth is enabled here (it is easy to do).
```

> We use TLS(with server certificate verification) only for external traffic(ingress, swift). Which seems sufficient

Reviewer: ```suggestion
We use TLS (with server certificate verification) only for external traffic (ingress, swift). Which seems sufficient
```

Also, usually flux2 resources are used for e.g. Helm, so flux2 controllers need to be installed in any destination cluster.

## Installation

### Prerequisites

#### Fluxcd
```
curl -s https://fluxcd.io/install.sh | sudo FLUX_VERSION=0.40.2 bash
flux install
```

### Install and wait for operators
```
$ k apply -k operators/
$ flux get helmrelease -n default
NAME                    REVISION        SUSPENDED       READY   MESSAGE
cert-manager            v1.11.0         False           True    Release reconciliation succeeded
ingress-nginx           4.5.2           False           True    Release reconciliation succeeded
postgres-operator       1.9.0           False           True    Release reconciliation succeeded
redis-operator          3.2.7           False           True    Release reconciliation succeeded
```

> Install separate operator by  e.g. `k apply -k operators/postgres/`

Reviewer 2: ```suggestion
Install separate operator by e.g. `k apply -k operators/postgres/`
```

Just a superfluous whitespace

Author: Thanks for your comments, I fix them in the 503fe3777ed62ba19822a4b32862cb95db152302, where I also little bit reorganized the repository structure and did some other changes, e.g. kustomize patches.

### Create redis and postgres cluster
```
bash envs/public/redis-secret.bash
k apply -k redis/
k apply -k postgres/
```

### Install Harbor

Take *ingress-nginx-controller* loadbalancer IP address and create DNS record for Harbor.

Generate secrets and install Harbor:
```
bash envs/public/harbor-secrets.bash
bash envs/public/swift-secret.bash <username> <password>
k apply -k envs/public/
```

### All in one installation using fluxcd Kustomization and GitRepository reconciliation

```
bash envs/public/redis-secret.bash
bash envs/public/harbor-secrets.bash

> #bash envs/public/swift-secret.bash <username> <password>

Reviewer 2: Is this a leftover or commented out for clarifying reasons?
